### PR TITLE
PR: Set projet filenames correctly when closing a project

### LINKED ISSUE #8375: Spyder editor does not open previously opened files in a project
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
Spyder editor does not open previously opened files in a project.

At any time, I may switch between projects. I would close project A, and open project B. But if I want to return to work on project A, I find that the editor does not automatically open all the files that was opened before I closed project A the last time.

Only when I close Spyder entirely, then will the project state be saved.

### What steps reproduce the problem?

1. Install anaconda
2. Open spyder
3. Create a project with files in them. 
4. Open the project, open some saved file Z in editor. Close all other files.
5. Close project.
6. Open project again.
7. Check if file Z is opened in editor.

### What is the expected output? What do you see instead?
File Z should be opened. It is not opened.

* Spyder version: 3.3.1
* Python version: 3.5.4 64 bits
* Qt version: 5.6.2
* PyQt version: 5.6
* Operating System name/version: Windows 8.1

### PR DESCRIPTION
**Pull request by jnsebgosselin**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

When closing a project, call `set_project_filenames` before setting the current project to `None`.

### Issue(s) Resolved

Fixes #8375

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: Jean-Sébastien Gosselin

### CODE REVIEW

**Comment by jnsebgosselin** [comment]
@ccordoba12 Is this worth it to add a test for such a small fix? Would you?

**Comment by jnsebgosselin** [comment]
Forget my last comment, I found a way to cover this at the plugin level without too much effort.

File: spyder/plugins/projects/tests/test_plugin.py [hunk 1]
```diff
@@ -120,5 +127,33 @@ def test_open_project_uses_visible_config(projects, tmpdir, value):
...
+
+    Regression test for Issue #8375
+    """
+    opened_files = ['file1', 'file2', 'file3']
+    mocker.patch.object(
+        projects.main.editor, 'get_open_filenames', return_value=opened_files)
+
+    # We mock os.path.isfile so that we do not have to
+    # actually create the files on the disk.
+    mocker.patch('spyder.plugins.projects.api.os.path.isfile',
```
Comment (CAM-Gerlach) [inline spyder/plugins/projects/tests/test_plugin.py]: Is this what we're using now instead of ``monkeypatch``?

File: spyder/plugins/projects/tests/test_plugin.py [hunk 2]
```diff
@@ -25,16 +25,25 @@
 from spyder.py3compat import to_text_string
 
 
-# =============================================================================
-# Fixtures
-# =============================================================================
+# ---- Fixtures
```
Comment (CAM-Gerlach) [inline spyder/plugins/projects/tests/test_plugin.py]: ??? I thought this was our standard test file formatting for new and substantially revised files? I've never seen the ``# ---- NAME`` formatting anywhere else in our code (or anywhere else that I've seen) while it seems the previous is much more conventional (and besides, looks way better and clearer IMO).

File: spyder/plugins/projects/tests/test_plugin.py [hunk 3]
```diff
@@ -57,9 +66,7 @@ def __getattr__(self, attr):
     return projects
 
 
-# =============================================================================
-# Tests
-# =============================================================================
+# ---- Tests
```
Comment (CAM-Gerlach) [inline spyder/plugins/projects/tests/test_plugin.py]: Same as above

File: spyder/plugins/projects/tests/test_plugin.py [hunk 4]
```diff
@@ -120,5 +127,33 @@ def test_open_project_uses_visible_config(projects, tmpdir, value):
...
+    mocker.patch('spyder.plugins.projects.api.os.path.isfile',
+                 return_value=True)
+
+    # Create the project.
+    path = to_text_string(tmpdir.mkdir('project1'))
+    projects.open_project(path=path)
+    assert projects.get_project_filenames() == []
+
+    # Close and reopen the project.
+    projects.close_project()
```
Comment (CAM-Gerlach) [inline spyder/plugins/projects/tests/test_plugin.py]: Although the original bug only occurred on closing and re-opening projects, would it be too much work to test switching projects as well in case there's a regression or edge case in the future?

File: spyder/plugins/projects/tests/test_plugin.py (hunk 2 above)
Comment (ccordoba12) [reply spyder/plugins/projects/tests/test_plugin.py]: I agree with @CAM-Gerlach here. Please revert.

File: spyder/plugins/projects/tests/test_plugin.py (hunk 4 above)
Comment (ccordoba12) [reply spyder/plugins/projects/tests/test_plugin.py]: I agree with this too. It shouldn't be that hard.

File: spyder/plugins/projects/tests/test_plugin.py (hunk 2 above)
Comment (jnsebgosselin) [reply spyder/plugins/projects/tests/test_plugin.py]: !!! The reason I did this was because special comments are added to the Outline Explorer, which I find very useful to separate at a glance the fixtures from the tests. But that's ok, I'm going to revert it.

image

Comment (ccordoba12) [reply spyder/plugins/projects/tests/test_plugin.py]: Sorry, then I agree with your change. Please remove your last commit.

File: spyder/plugins/projects/tests/test_plugin.py (hunk 4 above)
Comment (jnsebgosselin) [reply spyder/plugins/projects/tests/test_plugin.py]: This is a good idea. I will add that.

**Comment by ccordoba12** [comment]
@jnsebgosselin, you need to rebase/merge with master to fix the tests in Azure.

File: spyder/plugins/projects/tests/test_plugin.py (hunk 1 above)
Comment (CAM-Gerlach) [reply spyder/plugins/projects/tests/test_plugin.py]: I don't know, that was just my understanding from the previous files but its up to @ccordoba12

File: spyder/plugins/projects/tests/test_plugin.py (hunk 2 above)
Comment (CAM-Gerlach) [reply spyder/plugins/projects/tests/test_plugin.py]: What about

```python
# =============================================================================
# ---- Fixtures
# =============================================================================
```

That keeps most of the existing convention (and more visible/better looking styling) and only requires adding four characters on one line to existing files as they are modified, and is more visible inline without the outline explorer while still achieving the same affect with it.

Comment (jnsebgosselin) [reply spyder/plugins/projects/tests/test_plugin.py]: 👍 👍 👍 I think this is a good idea. If @ccordoba12 agrees with it I will do that.

Comment (ccordoba12) [reply spyder/plugins/projects/tests/test_plugin.py]: Sure, I agree with that too.

File: spyder/plugins/projects/tests/test_plugin.py (hunk 1 above)
Comment (ccordoba12) [reply spyder/plugins/projects/tests/test_plugin.py]: I don't have a problem with using any approach.

**Comment by jnsebgosselin** [comment]
@ccordoba12 I added the requested additional test. There is a chance you won't like my `create_projects` fixture that I've created to remove some repetitiveness of the code in both tests I've added.

There is no problem if you don't like it, I will just remove the 2 latest commits.

File: spyder/plugins/projects/tests/test_plugin.py [hunk 5]
```diff
@@ -120,5 +156,49 @@ def test_open_project_uses_visible_config(projects, tmpdir, value):
...
+    then reopened.
+
+    Regression test for Issue #8375
+    """
+    opened_files = ['file1', 'file2', 'file3']
+    path = to_text_string(tmpdir.mkdir('project1'))
+
+    # Create the projects plugin.
+    projects = create_projects(path, opened_files)
+    assert projects.get_project_filenames() == []
```
Comment (ccordoba12) [inline spyder/plugins/projects/tests/test_plugin.py]: Why is this empty and not equal to `opened_files`?

Comment (jnsebgosselin) [reply spyder/plugins/projects/tests/test_plugin.py]: `get_project_filenames` returns what is saved in the project's config file and this gets updated only when closing the project, not as new files are opened or closed in the Editor.

Comment (jnsebgosselin) [reply spyder/plugins/projects/tests/test_plugin.py]: So `create_project` open a new project with an empty list of files at first in the config. Then when we close the project, the list of files we are mocking as open in the Editor are set as the project filenames in the config file.

Comment (ccordoba12) [reply spyder/plugins/projects/tests/test_plugin.py]: Ok, I see. Thanks for the explanation!
